### ISSUE TITLE
[Q] Possible to use the websocket endpoint? 

### ISSUE THREAD

**Issue by sigaloid** [austinhuang0131/barinsta#653]
From what I can tell, the DM pull request doesn't implement this. This seems like a lot of reverse engineering the responses and requests to see exactly what structure they are supposed to be. On the webpage on firefox, it's possible to inspect the websocket requests back and forth, see which channels need to be "subscribed to" on page load, etc. This would allow real time seen, typing, and message indicators. It's also the official endpoint that the website uses, and wouldn't require hitting endpoints repeatedly on each refresh; it could just take the data from the websocket pubsub. 

wss://edge-chat.instagram.com/chat

**Comment by austinhuang0131** [comment]
we haven't reached that part yet lol

**Comment by sigaloid** [comment]
Understandable. I'm gonna try to gather info on that endpoint and its functionality. Would you be interested in me posting it here?

**Comment by austinhuang0131** [comment]
sure

**Comment by austinhuang0131** [comment]
but also, the app uses `edge-mqtt.facebook.com` for websocket, so there's significant difference between app websocket and browser websocket

**Comment by sigaloid** [comment]
I can't even get any websocket clients to connect to the endpoint. It either fails on TLS handshake, redirects, or it looks like it goes through just fine but the server says nothing and drops connection, seemingly before I can even send an initial message. I can't tell if this is some strange anti-reversal or if I'm doing something wrong. But I have tried almost everything, custom headers, setting cookie, etc. I even made the initial connection identical headers to the one firefox makes, confirmed by wireshark that they are identical connection requests. Either way I agree with your initial comment :cry:

**Comment by austinhuang0131** [comment]
Or you can try to study [this library](https://github.com/Nerixyz/instagram_mqtt)?

**Comment by sigaloid** [comment]
swag ill look into it

**Comment by austinhuang0131** [comment]
Any progress?

**Comment by sigaloid** [comment]
There's something different that library does, and I haven't been able to track it down. Websocket connection gets closed before messages can even send. It may be the library I'm using. It's far more complicated than it looks and the browser reversal is not going well. 

I've attached the right headers, the right auth tokens, etc and my best guess is there's something the websocket js on the webpage does that is not within ws spec. And trying to connect with a standard WS lib isn't working. 

In addition: https://news.ycombinator.com/item?id=22209892

**Comment by austinhuang0131** [comment]
No, they threw away the playbook (app websocket, for example, uses "thrift payload" to authenticate), so forget standard ws/mqtt spec.

Yes, I'm well aware about mgp25, but the effect is marginal.

**Comment by sigaloid** [comment]
I found [this library](https://github.com/tulir/mautrix-instagram) to be easier to comprehend than the lib you linked. It's a matrix/insta bridge in python. For example https://github.com/tulir/mautrix-instagram/blob/master/mauigpapi/mqtt/conn.py 

Keep in mind I don't know java, best I can do is pseudocode because I cannot get the connection to open through websockest (I have tried everything and I am losing my mind). I see you made a repo for MQTT, I wish you luck because either it's difficult to reverse or I'm skipping something incredibly basic and elementary.
